toggler/services/config.py

```diff
@@ -0,0 +1,80 @@
+"""Utilities to prepare configuration data for toggler."""
+
+import os
+import yaml
+import mergedeep
+from typing import MutableMapping
+
+from ..types import OptionalStream, OptionalStreams, OptionalPath, OptionalPaths
+from ..model import Feature
+from .env import ENV_KEY_CFG
+
+MERGE_STRAT = mergedeep.Strategy.REPLACE
+
+
+def prepare_config(
+    mode: str, streams: OptionalStreams = None, paths: OptionalPaths = None
+):
+    """Prepare config to be used for toggler.
+
+    1. Configurations are loaded from streams or paths.
+    2. If there are multiple configs, they are merged into one. Order
+        is important. Last in order configs overwrite data.
+    3. Merged raw configuration is parsed into configuration with features
+        activity per specified mode.
+    """
+    configs = load_configs(streams, paths)
+    config = merge_configs(configs)
+    return parse_config(mode, config)
+
+
+def load_configs(
+    streams: OptionalStreams = None, paths: OptionalPaths = None
+) -> list[dict]:
+    configs = []
+    args = _prepare_multi_load_cfg_args(streams, paths)
+    for kwargs in args:
+        configs.append(_load_config(**kwargs))
+    return configs
+
+
+def merge_configs(configs: list[dict]) -> MutableMapping:
+    if not configs:
+        return {}
+    main_cfg, other_configs = configs[0], configs[1:]
+    return mergedeep.merge(main_cfg, *other_configs, strategy=MERGE_STRAT)
+
+
+def parse_config(mode: str, config: dict | MutableMapping) -> dict[str, Feature]:
+    cfg = {}
+    for name, feature_data in config.items():
+        cfg[name] = Feature.from_feature_dict(mode, name, feature_data)
+    return cfg
+
+
+def _prepare_multi_load_cfg_args(
+    streams: OptionalStreams = None, paths: OptionalPaths = None
+) -> list[dict]:
+    if streams:
+        return [{"stream": stream} for stream in streams]
+    if paths:
+        return [{"path": path} for path in paths]
+    path_cfg = os.environ.get(ENV_KEY_CFG)
+    if path_cfg:
+        return [{"path": p.strip()} for p in path_cfg.split(",")]
+    raise ValueError(
+        f"To initiate Toggler Configuration, either stream, path or {ENV_KEY_CFG}"
```

Comment:
Wondering if this + is needed there? Maybe use parentheses instead? Or you think using + its more readable?

Reply:
What do you mean parenthesis?  String does not fit, so you have to split it into multiple lines (this is not the same as multiline string as that one would produce `\n` on each line). And yes `+` is to avoid ambiguity. Problem with implicit concatenation, is that it can easily used differently than expected. For example, you might be passing some string arguments and would forget comma between them, so instead of getting any error, you would get one argument concatenated. You would get misleading error. So doing explicit concat with `+`, we are showing that are actually really doing that:).

There was even a pep to remove implicit concat at all, but it was rejected (even though saying it was good suggestion, but it was too intrusive to change at this point): https://peps.python.org/pep-3126/